hudi-common/src/main/java/org/apache/hudi/common/table/read/HoodieBaseFileGroupRecordBuffer.java

```diff
@@ -126,12 +128,13 @@ protected Option<T> doProcessNextDataRecord(T record,
       // Merge and store the combined record
       // Note that the incoming `record` is from an older commit, so it should be put as
       // the `older` in the merge API
+
       HoodieRecord<T> combinedRecord = (HoodieRecord<T>) recordMerger.merge(
-          readerContext.constructHoodieRecord(Option.of(record), metadata, readerSchema),
-          readerSchema,
+          readerContext.constructHoodieRecord(Option.of(record), metadata),
+          (Schema) metadata.get(INTERNAL_META_SCHEMA),
           readerContext.constructHoodieRecord(
-              existingRecordMetadataPair.getLeft(), existingRecordMetadataPair.getRight(), readerSchema),
-          readerSchema,
+              existingRecordMetadataPair.getLeft(), existingRecordMetadataPair.getRight()),
+          (Schema) existingRecordMetadataPair.getRight().get(INTERNAL_META_SCHEMA),
           payloadProps).get().getLeft();
```

Comment:
Why not just initialize the `readerSchema` with `(Schema) metadata.get(INTERNAL_META_SCHEMA)` ?

Reply:
When there are more log files, partial updates, and schema evolution, `(Schema) metadata.get(INTERNAL_META_SCHEMA)` can be different across record keys.

Reply:
But it is specific per-file at lest right?

Reply:
To clarify, for reading log files, the reader schema is fetched from the header.  Here we're doing record-level merging.  Depending the log file from which the records come, the schema could be different.  However, the reference to the schema is the same as the schema instance is passed from the log reader.